core/commands/dag/dag.go

```diff
@@ -241,3 +279,435 @@ var DagResolveCmd = &cmds.Command{
+			nd, err := ipld.Decode(block)
+			if err != nil {
+				ret <- importResult{err: err}
+				return
+			}
+
+			if err := batch.Add(req.Context, nd); err != nil {
+				ret <- importResult{err: err}
+				return
+			}
+
+			counts.Blocks++
+
+			// encountered something known to be a root, for the first time
+			if seen, exists := roots[nd.Cid()]; exists && !seen {
+				roots[nd.Cid()] = true
+				counts.RootsSeen++
+			}
+		}
+
+		// every single file in it. is already open before we start
+		// just close here sooner rather than later for neatness
+		// this won't/can't help with not running out of handles
+		file.Close()
+	}
+
+	if err := it.Err(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	if err := batch.Commit(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	ret <- importResult{roots: &roots}
+}
+
+var DagExportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Streams the selected DAG as a .car stream on stdout.",
+		ShortDescription: `
+'ipfs dag export' fetches a dag and streams it out as a well-formed .car file.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("selector", true, false, "Selector representing the dag to export").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI.").WithDefault(true),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		c, err := cid.Decode(req.Arguments[0])
+		if err != nil {
+			return fmt.Errorf(
+				"unable to parse selector (currently only bare CIDs are supported): %s",
+				err,
+			)
+		}
+
+		// The current interface of go-car is rather suboptimal as it
+		// only takes a blockstore, instead of accepting a dagservice,
+		// and leveraging parallel-fetch capabilities
+		//
+		// Until the above is fixed, pre-warm the blockstore before doing
+		// anything else. We explicitly *DO NOT* take a lock during this
+		// operation: even if we lose some of the blocks we just received
+		// due to a conflicting GC: we will just re-retrieve everything when
+		// the car is being streamed out
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		if err := mdag.FetchGraph(req.Context, c, node.DAG); err != nil {
+			if !node.IsOnline {
+				err = fmt.Errorf("%s (currently offline, perhaps retry after attaching to the network)", err)
+			}
+			return err
+		}
+
+		sess := blockservice.NewSession(
+			req.Context,
+			node.Blocks,
+		)
+
+		var nodeCount int64
+
+		// The second part of the above - make a super-thin wrapper around
+		// a blockservice session, translating Session.GetBlock() to Blockstore.Get()
+		// The function also doubles as our progress meter
+		var wrapper getBlockFromSessionWrapper = func(c cid.Cid) (blk.Block, error) {
+			nodeCount++
+			return sess.GetBlock(req.Context, c)
+		}
+		sb := gipselectorbuilder.NewSelectorSpecBuilder(gipfree.NodeBuilder())
+		car := gocar.NewSelectiveCar(
```

Comment:
Do we care about the case where you don't have a complete graph available? I believe that the `*` selector is going to result in an error if it reaches an edge that `sess.GetBlock()` can't return for a CID. We just can't currently disambiguate the error type from ipld-prime to know whether that may be acceptable (e.g. if a `--allow-incomplete` type flag was introduced), but we could probably change it to get one if that's needed. (https://github.com/ipld/go-ipld-prime/blob/eb71617f4aeb7e73f36276f14e76e47fd90241ef/traversal/walk.go#L119)

Reply:
@rvagg on export we want to either output the complete graph ( potentially grabbing it from the network ), or to fail. I.e. at present we do not at all have `--allow-incomplete` in spirit, so that the .car files *produced* by go-ipfs are exclusively "well formed and deterministic" ( deterministic as dictated by go-ipld-prime ).

The import of course is a complete free-for-all ( also by design ) as can be seen here:
https://github.com/ipfs/go-ipfs/blob/03fe9bde87462aa11775bf6e8361a56fd967f89d/test/sharness/t0054-dag-car-import-export.sh#L41-L55
based on the datasets: https://github.com/ipfs/go-ipfs/blob/03fe9bde87462aa11775bf6e8361a56fd967f89d/test/sharness/t0054-dag-car-import-export-data/README.md

Let me know if this is potentially problematic...

Reply:
Comment above, it's not necessarily problematic, it just opens up API feature space that may be difficult to claw back in the future if we decide to be more strict. My personal inclination with such things these days is to be explicit and strict and open up feature space only in response to requests and demonstrated need (it also gives you a much more explicit space to construct tests around). Not my call here, so just a word of caution.